### PR TITLE
KEYCLOAK-16729 Add more info for how to retrieve provider types

### PR DESCRIPTION
**Pull request by mposolda**
@andymunro Do you please have a chance to review?

### CODE REVIEW

File: server_development/topics/providers.adoc [hunk 1]
```diff
@@ -153,6 +154,36 @@ public class MyThemeSelectorProviderFactory implements ThemeSelectorProviderFact
 }
 ----
 
+[[_use_available_providers]]
+=== Use available providers
+
+In your provider implementation, you can use other providers available in the {project_name}. The existing providers can be typically retrieved with the
```
Comment (andymunro) [inline server_development/topics/providers.adoc]: remove "the" before " {project_name}"

File: server_development/topics/providers.adoc [hunk 2]
```diff
@@ -153,6 +154,36 @@ public class MyThemeSelectorProviderFactory implements ThemeSelectorProviderFact
...
+usage of the `KeycloakSession`, which is available to your provider as described <<_implementing_spi,above>>.
```
Comment (andymunro) [inline server_development/topics/providers.adoc]: described in

File: server_development/topics/providers.adoc [hunk 3]
```diff
@@ -153,6 +154,36 @@ public class MyThemeSelectorProviderFactory implements ThemeSelectorProviderFact
...
+
+{project_name} has two provider types:
+
+* *Single-implementation provider types* - There can be only single active implementation of the particular provider type in the {project_name} runtime.
```
Comment (andymunro) [inline server_development/topics/providers.adoc]: remove "the" before " {project_name}"

File: server_development/topics/providers.adoc [hunk 4]
```diff
@@ -153,6 +154,36 @@ public class MyThemeSelectorProviderFactory implements ThemeSelectorProviderFact
...
+For example `HostnameProvider` specifies the hostname to be used by {project_name} and that is shared for the whole {project_name} server.
+Hence there can be only single implementation of this provider active for the {project_name} server. If there are multiple provider implementations available to the server runtime,
```
Comment (andymunro) [inline server_development/topics/providers.adoc:167]: add "a" before "single"

File: server_development/topics/providers.adoc [hunk 5]
```diff
@@ -153,6 +154,36 @@ public class MyThemeSelectorProviderFactory implements ThemeSelectorProviderFact
...
+[source,xml]
+----
+<spi name="hostname">
+    <default-provider>default</default-provider>
+    ...
+</spi>
+----
+In the code, you can obtain the provider like this `keycloakSession.getProvider(HostnameProvider.class)`
+
+* *Multiple implementation provider types* - Those are provider types, which allow to have multiple implementations available and working together
```
Comment (andymunro) [inline server_development/topics/providers.adoc]: Change ", which allow to have" to "that allow"

File: server_development/topics/providers.adoc [hunk 6]
```diff
@@ -153,6 +154,36 @@ public class MyThemeSelectorProviderFactory implements ThemeSelectorProviderFact
...
+in the {project_name} runtime. For example `EventListener` provider allows to have multiple implementations available and registered, which means
+that particular event can be sent to all the listeners (jboss-logging, sysout etc). In the code, you can obtain specified instance of the provider
```
Comment (andymunro) [inline server_development/topics/providers.adoc]: a specified

File: server_development/topics/providers.adoc [hunk 7]
```diff
@@ -153,6 +154,36 @@ public class MyThemeSelectorProviderFactory implements ThemeSelectorProviderFact
...
+for example like `session.getProvider(EventListener.class, "jboss-logging")` . You need to specify `provider_id` of the provider as the second argument
```
Comment (andymunro) [inline server_development/topics/providers.adoc]: chnage "like" to "such as"   -- in all cases

Add "the" after "specify"

**Review by andymunro — CHANGES_REQUESTED** [review]
Looks good.  Minor changes, Marek

**Comment by mposolda** [comment]
@andymunro I've applied all your suggestions. Thanks!

**Pull request merged**